arm11/source/start.s

```diff
@@ -127,4 +142,19 @@ wait_arm9:
 	@ Jump to the kernel!
 	bx lr
 
+smp_boot:
+	@ r0 is the current CPU ID
+	ldr r1, =SYNC_ADDR
+	add r1, r1, r0, lsl #2
+	mov r12, #0
+	str r12, [r1]
+
+wait_smpboot:
+	ldr r12, [r1]
+	cmp r12, #0
+	wfeeq
+	beq wait_smpboot
```

Comment:
If I make this unconditional, I can boot again. So it seems like the address we're getting is wrong?

Reply:
In dmesg I see a warning splat:
```
DT /cpu 3 nodes greater than max cores 2. capping them
...
Hardware name: Nintendo 3DS/CTR (Device Tree)
...
<stack trace from arm_dt_init_cpu_maps called from setup_arch called from start_kernel>
```
It's just a warning, not a panic, but perhaps we shouldn't be setting nr_cpus in the bootargs or something.

Reply:
I might have set the defconfig to only use 2 cores at maximum, that should be changed to 4.

Also do you mean making only `wfeeq` unconditional? If that works I suppose it's acceptable since the other CPUs will only be spinning for very short amounts of time.

@xerpi has also suggested changing the sev() to dsb_sev() in the kernel plat_smp.c, could you try that with the conditional `wfe`?

Reply:
ah this splat coming from the config!

```sh
$ grep NR_CPUS arch/arm/configs/nintendo3ds_defconfig 
CONFIG_NR_CPUS=2
```

Reply:
> Also do you mean making only wfeeq unconditional?

Ah, sorry, the branch `b wait_smpboot` so an inf loop.

> @xerpi has also suggested changing the sev() to dsb_sev() in the kernel plat_smp.c, could you try that with the conditional wfe?

No luck with `dsb_sev()` instead of `sev()` in arch/arm/mach-ctr/platsmp.c.

Reply:
Also, I think somethings going wrong with the address calculation, or where we're reading it from (wrong address, perhaps)? Since the crash is when jumping to the address pointed to by `r12`.  `dsb_sev` vs `sev` would just get us out of the `wfe` based loop, which we're definitely exiting.

https://community.arm.com/developer/tools-software/tools/f/armds-forum/940/synchronization-in-smp-system-wfe-sev-dmb-dsb

Reply:
yeah as noted in that other discussion the dsb is always happening because `iounmap` implicitly does a `dsb` before sending the event
I still don't fully understand how it's possible that it crashes, everything should work fine.
Try it with a vanilla staging branch rather than one with your changes in?

Reply:
Same results, even went back to devkitARM.

Reply:
So [before](https://github.com/linux-3ds/linux/commit/b3995dd4b5c2ef52f7eac6fddc7f2b5b0c415b2f), we were making core 1 start at `ctr_secondary_startup`.  Now you're making it start at `_start`. `_start` runs all kinds of additional stuff and the `mcr`'s at the beginning of the two look quite different. Could any of that be related?

Reply:
ha! That was it!
```diff
diff --git a/arm11/source/start.s b/arm11/source/start.s
index cd7e5a4..61be0ab 100644
--- a/arm11/source/start.s
+++ b/arm11/source/start.s
@@ -58,6 +58,7 @@ _start:
        @ Clear exclusive records
        clrex
 
+smp_start:
        @ Get the CPUID
        @ Branch to secondary boot if its not CPU0
        mrc p15, 0, r0, c0, c0, 5
@@ -86,7 +87,7 @@ wait_arm9:
 
        @ Trigger secondary CPU
        ldr r3, =0x1FFFFFDC
-       adr r4, _start
+       adr r4, smp_start
        str r4, [r3]
 
        ldr r3, =(SCU_BASE_ADDR + 0x1F00)
```

Reply:
that's weird and I'm not even sure if correct

Booting straight into smp_bolt might be dangerous because r0 is not necessarily the cpuid

Reply:
err...please double check. I think you may have confused my newly added label, `smp_start` which is just before fetching the cpuid via `mrc` with `ctr_secondary_startup` (or something else).  Otherwise you're comment doesn't make sense.

Reply:
whoops, yeah that looks right now that you mention it

Reply:
okay, changed it in my latest commit